ddtrace/settings/matching.py

```diff
@@ -0,0 +1,123 @@
+import os
+from typing import Dict
+from typing import Optional
+
+from ddtrace.internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+class SearchResult(object):
+    def __init__(self, match, cost):
+        # type: (str, int) -> None
+        self.match = match
+        self.cost = cost
+
+    def __str__(self):
+        return "<SearchResult(match=%s,cost=%s)>" % (
+            self.match,
+            self.cost,
+        )
+
+
+def min_cost_match(a, b):
+    # type: (Optional[SearchResult], Optional[SearchResult]) -> Optional[SearchResult]
+    if a is None:
+        return b
+    elif b is None:
+        return a
+    if a.cost > b.cost:
+        return b
+    return a
+
+
+class Trie(object):
```

Comment:
Can we leverage the `difflib` module here somehow?

Reply:
I've taken a look at the `difflib` module, and it's based on another algorithm (substring matching).
I think the one currently implemented works better to correct typos for two reasons:
* The algorithm difflib uses returns a score that depends on the length of the strings matched, whereas we mostly care about the absolute number of differences, and it doesn't consider character swap as a single change.  
* The current implementation should be more efficient on long strings because we can prune results early by giving a max distance of 1 or 2. That makes the matching mostly O(len(text)) whereas difflib is O(len(text) * len(pattern) * nb(pattern))